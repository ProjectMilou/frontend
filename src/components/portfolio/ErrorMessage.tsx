import React from 'react';
import { useTranslation } from 'react-i18next';
import {
  Avatar,
  Button,
  Card,
  CardActions,
  CardContent,
  CardHeader,
  makeStyles,
  Typography,
} from '@material-ui/core';
import { ErrorCode, errorMessageKey, errorTitleKey } from '../../Errors';

export type ErrorMessageProps = {
  messageKey?: string;
  error: ErrorCode;
  handling?: {
    buttonText: string;
    action: () => void;
  };
};

const useStyles = makeStyles({
  card: { maxWidth: '750px', margin: 'auto' },
  icon: { backgroundColor: '#B80C09' },
});

const ErrorMessage: React.FC<ErrorMessageProps> = ({
  messageKey,
  error,
  handling,
}) => {
  const { t } = useTranslation();
  const classes = useStyles();
  return (
    <Card className={classes.card}>
      <CardHeader
        avatar={<Avatar className={classes.icon}>!</Avatar>}
        title={
          <Typography variant="h5" component="h2">
            {t(errorTitleKey(error))}
          </Typography>
        }
        subheader={messageKey ? t(messageKey) : t('error.message.default')}
      />
      <CardContent>
        <Typography variant="body2" component="p">
          {t(errorMessageKey(error))}
        </Typography>
      </CardContent>
      {handling && (
        <CardActions>
          <Button size="small" onClick={handling.action}>
            {t(handling.buttonText)}
          </Button>
        </CardActions>
      )}
    </Card>
  );
};

export default ErrorMessage;
