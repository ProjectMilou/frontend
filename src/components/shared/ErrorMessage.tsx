import React from 'react';
import { useTranslation } from 'react-i18next';
import {
  Avatar,
  Button,
  Card,
  CardActions,
  CardContent,
  CardHeader,
  makeStyles,
  Typography,
} from '@material-ui/core';
import {
  errorMessageKey,
  errorTitleKey,
  isAuthenticationError,
} from '../../Errors';
import { Context } from '../../state/context';
import StorageService from '../../services/StorageService';

export type ErrorMessageProps = {
  /**
   * An optional translation key for a message that is displayed in addition to the error message.
   * Can be used to provide additional context to the user.
   */
  messageKey?: string;
  /** The error to display. */
  error: Error;
  /** Optional retry action. If set, a retry button is rendered. */
  retry?: () => void;
};

const useStyles = makeStyles({
  card: { maxWidth: '750px', margin: 'auto' },
  icon: { backgroundColor: '#B80C09' },
});

/**
 * Displays an error message with an optional retry button.
 *
 * * For authentication errors, the retry button opens the login dialog and
 *   removes the currently stored (invalid) token. After logging in
 *   successfully, the retry callback is called. Inside a {@link LoginWrapper},
 *   closing the dialog without logging in causes the LoginWrapper to navigate
 *   to the fallback URL.
 *
 * * For all other errors, the retry callback is called immediately on click.
 */
const ErrorMessage: React.FC<ErrorMessageProps> = ({
  messageKey,
  error,
  retry,
}) => {
  const { t } = useTranslation();
  const classes = useStyles();

  const { state, dispatch } = React.useContext(Context);
  const [loginOpened, setLoginOpened] = React.useState<boolean>(false);

  React.useEffect(() => {
    if (retry && loginOpened && state.loggedIn && !state.openLogin) {
      retry();
      setLoginOpened(false);
    }
  }, [retry, loginOpened, state]);

  return (
    <Card className={classes.card}>
      <CardHeader
        avatar={<Avatar className={classes.icon}>!</Avatar>}
        title={
          <Typography variant="h5" component="h2">
            {t(errorTitleKey(error))}
          </Typography>
        }
        subheader={messageKey ? t(messageKey) : t('error.message.default')}
      />
      <CardContent>
        <Typography variant="body2" component="p">
          {t(errorMessageKey(error))}
        </Typography>
      </CardContent>
      {retry && (
        <CardActions>
          {isAuthenticationError(error) ? (
            <Button
              size="small"
              onClick={() => {
                dispatch({ type: 'OPEN_LOGIN' });
                StorageService.removeToken();
                setLoginOpened(true);
              }}
            >
              {t('error.action.login')}
            </Button>
          ) : (
            <Button size="small" onClick={retry}>
              {t('error.action.retry')}
            </Button>
          )}
        </CardActions>
      )}
    </Card>
  );
};

export default ErrorMessage;
