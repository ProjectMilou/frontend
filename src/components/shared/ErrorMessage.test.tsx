import { render, fireEvent } from '@testing-library/react';
import * as React from 'react';
import ErrorMessage, { ErrorMessageProps } from './ErrorMessage';

describe('ErrorMessage', () => {
  afterEach(() => {
    jest.clearAllTimers();
  });

  const defaultProps: ErrorMessageProps = {
    error: 'UNKNOWN',
    handling: {
      buttonText: 'button',
      action: jest.fn(),
    },
    messageKey: 'message',
  };

  const renderComponent = (newProps?: Partial<ErrorMessageProps>) => {
    const props = { ...defaultProps, ...newProps };
    return {
      ...render(<ErrorMessage {...props} />),
      props,
    };
  };

  test('show and handle error message', () => {
    const { container, props, getByText } = renderComponent();
    expect(container).toMatchSnapshot();
    fireEvent.click(getByText(props.handling!.buttonText));
    expect(props.handling!.action).toHaveBeenCalled();
  });

  test('show error without message that cannot be handled', () => {
    const { container } = renderComponent({
      handling: undefined,
      messageKey: undefined,
    });
    expect(container).toMatchSnapshot();
  });
});
