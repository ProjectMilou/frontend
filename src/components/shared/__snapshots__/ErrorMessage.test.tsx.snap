// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage show and handle error message 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-icon-2 MuiAvatar-colorDefault"
        >
          !
        </div>
      </div>
      <div
        class="MuiCardHeader-content"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          error.title.UNKNOWN
        </h2>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          message
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        error.message.UNKNOWN
      </p>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          button
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`ErrorMessage show error without message that cannot be handled 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-icon-2 MuiAvatar-colorDefault"
        >
          !
        </div>
      </div>
      <div
        class="MuiCardHeader-content"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          error.title.UNKNOWN
        </h2>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          error.message.default
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        error.message.UNKNOWN
      </p>
    </div>
  </div>
</div>
`;
